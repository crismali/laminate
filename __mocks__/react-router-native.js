/* eslint-env jest */
const { Switch, Route, matchPath } = require.requireActual('react-router-native')
const withRouter = component => component

export { withRouter, Switch, Route, matchPath }
