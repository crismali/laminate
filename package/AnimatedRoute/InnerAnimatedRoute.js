import { Component } from 'react'
import { Animation } from '../Animation'

export class InnerAnimatedRoute extends Component {
  static propTypes = {
    ...Animation.childContextTypes
  }

  state = {
    children: this.props.children
  }

  componentDidMount () {
    if (this.props.children) {
      this.props.animation.play()
    }
  }

  componentWillReceiveProps (nextProps) {
    const { children: currentChildren } = this.props
    const { children: nextChildren } = nextProps

    if (this.isEntering(currentChildren, nextChildren)) {
      this.enter(nextChildren)
    } else if (this.isExiting(currentChildren, nextChildren)) {
      this.exit()
    } else {
      this.setState({ children: nextChildren })
    }
  }

  enter = (children) => {
    this.setState({ children }, this.props.animation.play)
  }

  exit = () => {
    this.props.animation.rewind({ callback: this.clearChildren })
  }

  clearChildren = () => this.setState({ children: null })

  isEntering = (currentChildren, nextChildren) => {
    return !currentChildren && nextChildren
  }

  isExiting = (currentChildren, nextChildren) => {
    return currentChildren && !nextChildren
  }

  render () {
    return this.state.children
  }
}
