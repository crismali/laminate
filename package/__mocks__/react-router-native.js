/* eslint-env jest */
import React from 'react'

const { Switch, Route, matchPath, StaticRouter } = require.requireActual('react-router-native')
const withRouter = component => component
const MemoryRouter = (props) => <StaticRouter context={{}} {...props} />

export { withRouter, Switch, Route, matchPath, MemoryRouter, StaticRouter }
