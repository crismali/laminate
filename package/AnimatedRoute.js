import React, { Component } from 'react'
import { Route } from 'react-router-native'
import { Animation } from './Animation'

class InnerAnimatedRoute extends Component {
  static propTypes = {
    ...Animation.childContextTypes
  }

  state = {
    children: this.props.children
  }

  componentDidMount () {
    if (this.props.children) {
      this.props.animation.play()
    }
  }

  componentWillReceiveProps (nextProps) {
    const { children: currentChildren } = this.props
    const { children: nextChildren } = nextProps

    if (this.isEntering(currentChildren, nextChildren)) {
      this.enter(nextChildren)
    } else if (this.isExiting(currentChildren, nextChildren)) {
      this.exit()
    } else {
      this.setState({ children: nextChildren })
    }
  }

  enter = (children) => {
    this.setState({ children }, this.props.animation.play)
  }

  exit = () => {
    this.props.animation.rewind({ callback: () => this.setState({ children: null }) })
  }

  isEntering = (currentChildren, nextChildren) => {
    return !currentChildren && nextChildren
  }

  isExiting = (currentChildren, nextChildren) => {
    return currentChildren && !nextChildren
  }

  render () {
    return this.state.children
  }
}

export class AnimatedRoute extends Component {
  static propTypes = {
    ...Route.propTypes
  }

  render () {
    const { component: GivenComponent, ...otherProps } = this.props

    return <Animation {...this.props}>
      {(animation) => (
        <Route {...otherProps}>
          {(routerProps) => (
            <InnerAnimatedRoute animation={animation}>
              {routerProps.match && <GivenComponent {...routerProps} animation={animation.value} />}
            </InnerAnimatedRoute>
          )}
        </Route>
      )}
    </Animation>
  }
}
