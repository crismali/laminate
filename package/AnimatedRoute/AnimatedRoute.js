import React, { Component } from 'react'
import { Route } from 'react-router-native'
import { Animation } from '../Animation'
import { InnerAnimatedRoute } from './InnerAnimatedRoute'

export class AnimatedRoute extends Component {
  static propTypes = {
    ...Route.propTypes
  }

  render () {
    const { component: GivenComponent, ...otherProps } = this.props

    return <Animation {...this.props}>
      {(animation) => (
        <Route {...otherProps}>
          {(routerProps) => (
            <InnerAnimatedRoute animation={animation}>
              {routerProps.match && <GivenComponent {...routerProps} animation={animation.value} />}
            </InnerAnimatedRoute>
          )}
        </Route>
      )}
    </Animation>
  }
}
