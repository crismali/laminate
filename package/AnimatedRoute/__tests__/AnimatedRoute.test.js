/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import { StaticRouter } from 'react-router-native'
import { AnimatedRoute } from '../AnimatedRoute'

const Child = () => <div>child</div>

jest.mock('../../Animation', () => {
  const { Animated } = require('react-native')
  const Animation = ({ children }) => children({ play: jest.fn(), rewind: jest.fn(), value: new Animated.Value(0) })
  return {
    Animation
  }
})

it('renders the given component when the route matches', () => {
  const subject = mount(<StaticRouter location='/foo' context={{}}>
    <AnimatedRoute path='/foo' component={Child} another='prop' />
  </StaticRouter>)
  expect(subject).toMatchSnapshot()
})

it('does not render the given component when the route does not match', () => {
  const subject = mount(<StaticRouter location='/bar' context={{}}>
    <AnimatedRoute path='/foo' component={Child} another='prop' />
  </StaticRouter>)
  expect(subject).toMatchSnapshot()
})
