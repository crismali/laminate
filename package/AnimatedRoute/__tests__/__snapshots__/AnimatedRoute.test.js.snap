// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render the given component when the route does not match 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/bar"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/bar",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AnimatedRoute
      another="prop"
      component={[Function]}
      path="/foo"
    >
      <Animation
        another="prop"
        component={[Function]}
        path="/foo"
      >
        <Route
          another="prop"
          path="/foo"
        >
          <InnerAnimatedRoute
            animation={
              Object {
                "play": [MockFunction],
                "rewind": [MockFunction],
                "value": Value {
                  "value": 0,
                },
              }
            }
          />
        </Route>
      </Animation>
    </AnimatedRoute>
  </Router>
</StaticRouter>
`;

exports[`renders the given component when the route matches 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/foo"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/foo",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AnimatedRoute
      another="prop"
      component={[Function]}
      path="/foo"
    >
      <Animation
        another="prop"
        component={[Function]}
        path="/foo"
      >
        <Route
          another="prop"
          path="/foo"
        >
          <InnerAnimatedRoute
            animation={
              Object {
                "play": [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                },
                "rewind": [MockFunction],
                "value": Value {
                  "value": 0,
                },
              }
            }
          >
            <Child
              animation={
                Value {
                  "value": 0,
                }
              }
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/foo",
                    "search": "",
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "pathname": "/foo",
                  "search": "",
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/foo",
                  "url": "/foo",
                }
              }
              staticContext={Object {}}
            >
              <div>
                child
              </div>
            </Child>
          </InnerAnimatedRoute>
        </Route>
      </Animation>
    </AnimatedRoute>
  </Router>
</StaticRouter>
`;
